pkg/pgclient/config.go

```diff
@@ -154,32 +158,76 @@ func (cfg *Config) GetConnectionStr() string {
+		log.Warn("msg", "configured db.connections-max is greater then postgres max_connections",
+			"db.connections-max", cfgMaxConns,
+			"max_connections", dbMaxConns,
+		)
 	}
 
-	switch {
-	case inputPoolSize == defaultPoolSize:
-		// For the default case, we need to take up defaultFraction of allowed connections.
-		poolSize := float64(maxConns) * defaultFraction
-		if cfg.UsesHA {
-			poolSize /= 2
+	calc := func(inputPoolSize int, fraction float64) int {
+		if inputPoolSize == defaultPoolSize {
+			poolSize := float64(dbMaxConns) * fraction
+			if cfg.UsesHA {
+				poolSize /= 2
+			}
+			return int(poolSize)
 		}
-		return int(poolSize), nil
-	case inputPoolSize < MinPoolSize:
-		return 0, fmt.Errorf("%s pool size canot be less than %d: received %d", poolName, MinPoolSize, inputPoolSize)
-	case inputPoolSize > maxConns:
-		return 0, fmt.Errorf("%s pool size canot be greater than the 'max_connections' allowed by the database", poolName)
-	default:
+		return inputPoolSize
+	}
+
+	// calc maint pool size
+	maintPoolSize = cfg.MaintPoolSize
+	if maintPoolSize == defaultPoolSize {
+		maintPoolSize = defaultMaintPoolSize
+	}
+
+	// calc writer pool size
+	if !readOnly {
+		writerPoolSize = calc(cfg.WriterPoolSize, defaultConnFraction)
+	}
+
+	// calc reader pool size
+	readerFraction := defaultReaderFraction
+	if readOnly {
+		readerFraction = defaultConnFraction
+	}
+	readerPoolSize = calc(cfg.ReaderPoolSize, readerFraction)
+	if cfg.ReaderPoolSize == defaultPoolSize && readerPoolSize+writerPoolSize+maintPoolSize > cfgMaxConns {
+		readerPoolSize = readerPoolSize - maintPoolSize
+	}
+
+	// check pool sizes against min pool size
+	if maintPoolSize < MinPoolSize {
+		return 0, 0, 0,
+			fmt.Errorf("maint pool size canot be less than %d: received %d", MinPoolSize, maintPoolSize)
+	}
+	if !readOnly && writerPoolSize < MinPoolSize {
+		return 0, 0, 0,
+			fmt.Errorf("writer pool size canot be less than %d: received %d", MinPoolSize, writerPoolSize)
+	}
+	if readerPoolSize < MinPoolSize {
+		return 0, 0, 0,
+			fmt.Errorf("reader pool size canot be less than %d: received %d", MinPoolSize, readerPoolSize)
+	}
+
+	// check sum of pool sizes against database max_connections
+	if readerPoolSize+writerPoolSize+maintPoolSize > dbMaxConns {
+		return 0, 0, 0,
+			fmt.Errorf("reader-pool (size=%d) + writer-pool (size=%d) + maint-pool (size=%d) more than postgres max_connections (%d). Decrease the pool-sizes", readerPoolSize, writerPoolSize, maintPoolSize, dbMaxConns)
 	}
-	if cfg.UsesHA {
-		inputPoolSize /= 2
+	// check sum of pool sizes against config db.connections-max
+	if readerPoolSize+writerPoolSize+maintPoolSize > cfgMaxConns {
+		return 0, 0, 0,
+			fmt.Errorf("reader-pool (size=%d) + writer-pool (size=%d) + maint-pool (size=%d) more than db.connections-max (%d). Increase the db.connections-max or decrease the pool-sizes", readerPoolSize, writerPoolSize, maintPoolSize, cfgMaxConns)
 	}
-	return inputPoolSize, nil
+	return
```

Comment:
Please no naked returns in such big functions.

Reply:
👍